### PR TITLE
[Workspace][Feature]Setup workspace skeleton and implement basic CRUD API

### LINKED ISSUE #4945: [Workspace][Feature] Setup workspace skeleton and implement basic CRUD API
**Feature includes**
- [x] Add feature flag on workspace plugin.
- [x] Setup workspace plugin skeleton.
- [x] Define data structure on `workspace`.
- [x] Implementation on CRUD operations on `workspace`.

**Additional context**

#4615

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

This PR is the initial PR for workspace, as we know workspace will be a new function that provides:
1. Organization on saved objects.
2. Manage visibility of opensearch features
3. Authorization on operations on saved objects.

And before all of the features mentioned above, we need to setup the workspace as a core plugin.

This PR includes:

- Add feature flag on workspace plugin.
- Setup workspace plugin skeleton.
- Define data structure on `workspace`.
- Implementation on CRUD operations on `workspace`.

### Issues Resolved

#4945 

## Screenshot

The changes introduced by this PR is not related to UI change.

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by SuZhou-Joe** [comment]
#4945

**Comment by SuZhou-Joe** [comment]
> Thanks for raising the initial PR for workspace feature! I noticed this is a draft, but two things I'd suggest to add:
> 
> 1. Add the configuration so that entire workspace plugin is disabled by default from the first workspace commit
> 2. Add detailed PR description

Sure, will add more test cases as well.

File: src/plugins/workspace/server/routes/index.ts
Comment (kavilla) [inline src/plugins/workspace/server/routes/index.ts:21]: I think it will be easier to start from or have some JSDOC friendly annotation here so that we can export it.

There a number of issues related to saved_objects APIs being confusing and not well documented. This feels like a wrapper to the saved objects so it will likely have the same complaints if there is a lack of documentation.

Also, allows for these APIs to SEMVER compliant

File: src/plugins/workspace/server/types.ts
Comment (kavilla) [inline src/plugins/workspace/server/types.ts:30]: whats the `db` stand for here?

File: src/plugins/workspace/server/plugin.ts
Comment (kavilla) [inline src/plugins/workspace/server/plugin.ts:28]: I believe somewhere you will want to register your saved object type of `workspace` so we can export that new saved object type if needed.

**Review by kavilla — COMMENTED** [review]
Thanks for opening this up!

A couple nits and couple of comments that I think should be considered. 
Specifically file whitespacing, custom random generate id to pass to the creating of the workspace instead of relying on OpenSearch, and documentation on the APIs that are updated to accurately reflect usage.

Also, something to consider when implemented the more complicated APIs, Saved Objects API supports `--compressed` so I'd imagine if I wanted to use this API I could use the same flag as well. Verified your stuff isn't impacted but just a headsup. Which can be extended on if we plan on workspaces being fully integrated with saved objects

File: src/plugins/workspace/public/plugin.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.ts:4]: Thanks and fixed, but if we want to make this consistent, maybe we can add some lint rules for that?

File: src/plugins/workspace/server/saved_objects/workspace.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace.ts:25]: Done.

File: test/api_integration/apis/workspace/index.ts
Comment (SuZhou-Joe) [reply test/api_integration/apis/workspace/index.ts:8]: Done, but if we want to remove the lodash, shall we add some lint rules first in case there are codes keep introducing Lodash usage.

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:73]: Done for this.

File: src/plugins/workspace/server/workspace_client.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/workspace_client.ts:25]: You are right, we thought there may be some WorkspaceClientWithSql or WorkspaceClientWithPostgress in the beginning. let me change that.

File: src/plugins/workspace/opensearch_dashboards.json
Comment (SuZhou-Joe) [reply src/plugins/workspace/opensearch_dashboards.json:6]: May not, workspace space relies on core.savedObjects, not plugins.savedObjects.

File: src/plugins/workspace/opensearch_dashboards.json
Comment (SuZhou-Joe) [reply src/plugins/workspace/opensearch_dashboards.json:5]: Thanks, set it to false

File: src/plugins/workspace/server/utils.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/utils.ts:11]: The workspaceId will be prepend to url in the following PR, so a uuid with a length of 32 will be a little bit too large for a url path or query, so we choose to make it shorter by using this method.

File: src/plugins/workspace/server/types.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/types.ts:30]: Change to `IWorkspaceClientImpl`

File: src/plugins/workspace/server/plugin.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts:29]: Thanks kavilla but we can export `workspace` in later PRs if it is needed, for this PR, we can keep as less exports as possbile.

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:38]: Oh, didn't know that. Do you mean that we should make `perPage` as a uiSettings that user can customize in advanced setttings?

But I think it is good to use some hard-coded default value for now because

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:21]: Hi kavilla, could you please give an example on how to "start from or have some JSDOC" and "SEMVER compliant". As far as I know, SEMVER compliant is used for version.

File: src/plugins/workspace/server/routes/index.ts
Comment (kavilla) [reply src/plugins/workspace/server/routes/index.ts:38]: Yeah! Which would be set within the public space of the code so yeah it's a good idea to keep this configure. It was more so a thing to add to your backlog.

File: src/plugins/workspace/server/utils.ts
Comment (kavilla) [reply src/plugins/workspace/server/utils.ts:11]: > The workspaceId will be prepended to url in the following PR, so a uuid with a length of 32 will be a little bit too large for a url path or query, so we choose to make it shorter by using this method.

That is pretty standard for the application. For example: 
https://playground.opensearch.org/app/dashboards#/view/7adfa750-4c81-11e8-b3d7-01146121b73d?_g=(filters:!(),refreshInterval:(pause:!f,value:900000),time:(from:now-24h,to:now))&_a=(description:'Analyze%20mock%20flight%20data%20for%20OpenSearch-Air,%20Logstash%20Airways,%20OpenSearch%20Dashboards%20Airlines%20and%20BeatsWest',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'%5BFlights%5D%20Global%20Flight%20Dashboard',viewMode:view)`.

The url is pretty large and is what we utilize for global and app state so this can get even longer. I would be worried about potential conflicts. Custom generate ID functions tend to not actually be as random as they imply so it would be a lot safer to defer to OpenSearch to index and provide you with a verified unique ID.

If the length of the URL can be shorten utilize the short URL function
Screenshot 2023-09-27 at 11 32 14 PM.

Or you can also make titles unique ids as well. I believe index patterns are the same way.

File: src/plugins/workspace/server/routes/index.ts
Comment (kavilla) [reply src/plugins/workspace/server/routes/index.ts:21]: For: https://semver.org/ 

> For this system to work, you first need to declare a public API. This may consist of documentation or be enforced by the code itself. Regardless, it is important that this API be clear and precise

So documentation can help with that (which can be inline for now).

Something like this: https://github.com/kavilla/opensearch-dashboards-sdk-js/blob/avillk/sdk_base_2_no_doc/lib/mount/mount_point.ts#L10

If it follows a specific format then we could add a library that can pull and generate markdown files automatically. For example, https://kavilla.github.io/opensearch-dashboards-sdk-js/modules/mount.html

File: src/plugins/workspace/server/types.ts
Comment (kavilla) [reply src/plugins/workspace/server/types.ts]: That helps a lot thank you!

File: src/plugins/workspace/server/plugin.ts
Comment (kavilla) [reply src/plugins/workspace/server/plugin.ts:29]: Gotcha, is there a meta issue or board that is keeping track of requirements?

I think we definitely still want to have the ability to export the workspace. Everything here https://playground.opensearch.org/app/management/opensearch-dashboards/objects is saved within the OSD index and exportable. If we export it out without workspaces but the other saved objects have the context of a workspace but the workspaces.

Comment (kavilla) [reply src/plugins/workspace/server/plugin.ts:29]: Sorry I realized that maybe I was little bit unclear with my first comment. So just in case, doing something like this part:

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_builder/server/plugin.ts#L29

File: src/plugins/workspace/opensearch_dashboards.json
Comment (kavilla) [inline src/plugins/workspace/opensearch_dashboards.json:7]: I think savedObjects is one I suggested. Just in case, I would suggest in a follow-up verifying all your requiredPlugins and bundles

**Comment by kavilla** [comment]
I think overall I think generally most of the comments can be a subsequent PRs as I imagine not having this one is a blocker to other PRs. Since it's disabled by default then it should be fine my biggest concern is the random gen id for 6 characters being used.

File: src/plugins/workspace/server/workspace_client.ts
Comment (kavilla) [reply src/plugins/workspace/server/workspace_client.ts]: Ah, if you are building with the intent of future state I believe we began started utilizing "Engine" (not that is was stated clearly for all to see). But within this PR: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4612/files

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:38]: Sure, will do that later.

File: src/plugins/workspace/server/plugin.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts:29]: Sure, we are doing the register [here](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5075/files#diff-38cb975b307e4ac766eed69f790e0312a0a08411a1d05960cfbbc9b951e6b99aR59)

File: src/plugins/workspace/server/utils.ts
Comment (kavilla) [reply src/plugins/workspace/server/utils.ts:11]: > (10000 / 256^6)

It's likely less than that:
https://nodejs.org/api/crypto.html#cryptorandombytessize-callback

> cryptographically strong pseudorandom data

So it will likely be based on a random seed value set by the OS.

> [https://playground.opensearch.org/w/${uuid_for_current_workspace}/app/dashboards],

Apologies, I wasn't aware of this. I believe this might have a larger implication to the application to how the application operates and will probably increase the scope of this project significantly.

Few examples on the top of my head:
* OSD starts up and discovers what plugins that are installed. It then takes the bundles for plugins and pushes under url/app/pluginName. So the current navigation system will have to be reworked. Because it will depend on url/app/pluginName to make the appropriate HTTP calls to download the bundles.
* History will have to be considered as well. Parts of the application and external plugins can be relying on the URL to handle history so if only parts of the application have the workspace route then they can get in a bad state.
* State management is managed by the URL via `_a` (for app) and `_g` so this might be impacted and you will have to modify existing OSD URL storage functions as some functions might be slicing up the route based on `url/{base_path}/app/plugin`
* APIs get hosted under url/api/pluginName

> make it more user friendly for user to see which routes they are viewing

Likely the URLs are obfuscated anyways depending on the size of your screen (mobile users). And if the goal is to be user friendly I think I would just go all the way and utilize the title as the unique id. 

**tl;dr**: So overall, I think that might add a lot of scope creep to add that and should be a proposal. To the point where I wouldn't want to consider that as a reason to keep in shorten Ids when we could add a new state like `_w`

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/utils.ts:11]: However things are difference between index patterns because workspace's title can be updated after a workspace has been created.

As for prepending workspaceId into our url path, there will be a RFC to discuss the options and we can forget that on this PR.

So what about we using a larger size, for example 12 maybe, as our workspaceId? As a length of 32 is too long for a url, especially when user copy a object detail page url, there will be a length of 32 to be used for workspaceId and a length of 32 to be used for object Id.

**Comment by xluo-aws** [comment]
> I think overall I think generally most of the comments can be a subsequent PRs as I imagine not having this one is a blocker to other PRs. Since it's disabled by default then it should be fine my biggest concern is the random gen id for 6 characters being used.

@kavilla , Hi, Rocky, wants to confirm if you still have big concern on 6 char workspaceid even we can automatically generate another id once detect there is a conflict? We have to check all workspaces to make sure the workspace name is unique so there is no overhead to compare workspace id at the same time.

**Comment by kavilla** [comment]
> > I think overall I think generally most of the comments can be a subsequent PRs as I imagine not having this one is a blocker to other PRs. Since it's disabled by default then it should be fine my biggest concern is the random gen id for 6 characters being used.
> 
> @kavilla , Hi, Rocky, wants to confirm if you still have big concern on 6 char workspaceid even we can automatically generate another id once detect there is a conflict? We have to check all workspaces to make sure the workspace name is unique so there is no overhead to compare workspace id at the same time. 

Yeah, i think the benefit to me is still lost to me. If for ux I think a better experience would be something similar to index patterns and utilizing the title as a unique key.

**Comment by xluo-aws** [comment]
> > > I think overall I think generally most of the comments can be a subsequent PRs as I imagine not having this one is a blocker to other PRs. Since it's disabled by default then it should be fine my biggest concern is the random gen id for 6 characters being used.
> > 
> > 
> > @kavilla , Hi, Rocky, wants to confirm if you still have big concern on 6 char workspaceid even we can automatically generate another id once detect there is a conflict? We have to check all workspaces to make sure the workspace name is unique so there is no overhead to compare workspace id at the same time.
> 
> Yeah, i think the benefit to me is still lost to me. If for ux I think a better experience would be something similar to index patterns and utilizing the title as a unique key. In the case of index patterns it passes the title as the id (I believe). For workspace, we could do a similar thing but allow for the title to be change independently from the id when editting.
> 
> This might not be accurate anymore, but from my understanding the id field autogenerated is based on Lucene and is not random as well. They are patterned out which assists with stuff like compression

@kavilla, what you said makes sense, but we're still not sure if that's the better approach, will need more discussion how to create the workspace id. We can create a RFC for this topic and solve it in 1-2 weeks. Can we merge this PR first so other PRs depending on this PR can be submitted and get reviewed soon?

**Comment by kavilla** [comment]
> > > > I think overall I think generally most of the comments can be a subsequent PRs as I imagine not having this one is a blocker to other PRs. Since it's disabled by default then it should be fine my biggest concern is the random gen id for 6 characters being used.
> > > 
> > > 
> > > @kavilla , Hi, Rocky, wants to confirm if you still have big concern on 6 char workspaceid even we can automatically generate another id once detect there is a conflict? We have to check all workspaces to make sure the workspace name is unique so there is no overhead to compare workspace id at the same time.
> > 
> > 
> > Yeah, i think the benefit to me is still lost to me. If for ux I think a better experience would be something similar to index patterns and utilizing the title as a unique key. In the case of index patterns it passes the title as the id (I believe). For workspace, we could do a similar thing but allow for the title to be change independently from the id when editting.
> > This might not be accurate anymore, but from my understanding the id field autogenerated is based on Lucene and is not random as well. They are patterned out which assists with stuff like compression
> 
> @kavilla, what you said makes sense, but we're still not sure if that's the better approach, will need more discussion how to create the workspace id. We can create a RFC for this topic and solve it in 1-2 weeks. 

sorry about the delay on responding. it got a little flooded. i think if you could call it a proposal? RFC seemed more finalized and if no comment then it's usually implemented.

> Can we merge this PR first so other PRs depending on this PR can be submitted and get reviewed soon?

yeah since it's not hooked into stuff right now I'm okay with that! But would you be able to create an issue and link it here?

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:21]: Done for this, and I can not find JSDoc on http routes, so I add JSDoc on workspaceClient, I think that will help as well.

**Comment by SuZhou-Joe** [comment]
> 
> yeah since it's not hooked into stuff right now I'm okay with that! But would you be able to create an issue and link it here?

Hi @kavilla , this issue is here #5201

**Review by kavilla — DISMISSED** [review]

File: src/plugins/workspace/server/integration_tests/routes.test.ts
Comment (joshuarrrr) [inline src/plugins/workspace/server/integration_tests/routes.test.ts:33]: Is there anyway to avoid the hard-coded delay?

File: src/plugins/workspace/server/integration_tests/routes.test.ts
Comment (joshuarrrr) [inline src/plugins/workspace/server/integration_tests/routes.test.ts:38]:
```suggestion
  describe('Workspace CRUD APIs', () => {
```

File: src/plugins/workspace/server/integration_tests/routes.test.ts
Comment (joshuarrrr) [inline src/plugins/workspace/server/integration_tests/routes.test.ts:143]: nit - a `list` test should generally test more than one result, as well as empty results.

**Review by joshuarrrr — COMMENTED** [review]
No significant concerns, although I agree with @kavilla that we probably want to come up with a better id or slug generation solution via https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5201.

File: src/plugins/workspace/server/utils.ts
Comment (zengyan-amazon) [reply src/plugins/workspace/server/utils.ts:11]: I don't think using a 6 character ID doesn't make big difference than a 32 bytes UUID. and agree with @kavilla that if you really want to make it friendly, you can use user specified friendly id as the workspace document id instead using auto generated ones.

And regarding prepending the workspace id in the url path, I think Kibana namespace does it, but do we compare other options like using query parameter?

**Review by zengyan-amazon — COMMENTED** [review]
we are making workspace as a core service on browser side anyways (https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5092), why making workspace management CRUD APIs in a plugin?

File: src/plugins/workspace/opensearch_dashboards.json
Comment (SuZhou-Joe) [reply src/plugins/workspace/opensearch_dashboards.json:5]: This is temporary, we will set `ui` to true once ui side changes are introduced.

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:33]: The reasons is that:
1. `Workspace` is a new concept and its metadata can be stored in any other storage, for now we store that as a type of saved objects but we should provide a layer that hides the implementation detail so that we can adjust with few changes.
2. There will be some extra logic to handle workspace object in the following PRs, like ACL format, validating duplication on workspace name or creating reserved workspace, it is better that we wrap those logic inside server side and make browser call as simple as possible.

File: src/plugins/workspace/server/plugin.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts:28]: Yes, the client now is a simple wrapper on top of savedObjectClient but in the following PRs more logic will be added into this client, and this PR is intended to setup the skeleton for them. And this client won't be too `heavy` as there will be only a single instance for each OSD server and we do not have any side effect inside the client.

File: src/plugins/workspace/public/plugin.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.ts:8]: You are right but we will have public side code changes in the following PRs and the entire workspace feature will be controlled behind a default-turned-off feature flag, can we keep it as following PRs rely on this skeleton?

File: src/plugins/workspace/server/integration_tests/routes.test.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/integration_tests/routes.test.ts]: Sure, I believed that I copied the delay from other integration_test cases but can not find it any more. Removed that.

File: src/core/types/workspace.ts
Comment (SuZhou-Joe) [reply src/core/types/workspace.ts]: The type is also used in #5092 , I just rebased main to discard the change to reduce confusion.

**Comment by SuZhou-Joe** [comment]
> but do we compare other options like using query parameter?
@zengyan-amazon , we did and there is a RFC(#5243 ) to compare all the options, including the one that append the id into query.

File: src/plugins/workspace/server/integration_tests/routes.test.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/integration_tests/routes.test.ts:157]: Updated.

File: test/api_integration/apis/workspace/index.ts
Comment (SuZhou-Joe) [reply test/api_integration/apis/workspace/index.ts:117]: Updated.

**Comment by SuZhou-Joe** [comment]
> we are making workspace as a core service on browser side anyways (#5092), why making workspace management CRUD APIs in a plugin?

core workspace service provides some Observables for other core or plugin to use, the CRUD methods and corresponding APIs should be controlled behind `workspace.enabled` feature flag so we implement those in a plugin.

**Comment by SuZhou-Joe** [comment]
@kavilla @joshuarrrr @zengyan-amazon , could we have two approval to make this PR merged?

**Review by kavilla — APPROVED** [review]

File: src/plugins/workspace/server/routes/index.ts
Comment (zengyan-amazon) [reply src/plugins/workspace/server/routes/index.ts:33]: my concern is that, even we have this client to deal with workspace related logic, since we model workspace as a type of saved object, users will still be able to use saved object APIs to access workspaces, while on those code path workspace operations doesn't have such logic in the workspace client. Maybe that can lead to un expected behaviors of workspace?

File: src/plugins/workspace/server/utils.ts
Comment (zengyan-amazon) [reply src/plugins/workspace/server/utils.ts:11]: @SuZhou-Joe any thinking around this?

File: src/plugins/workspace/server/routes/index.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/routes/index.ts:33]: Yeah, there are two ways that user can modify workspace metadata by bypassing workspace API:
1. Saved objects API provided by OSD: We add an ACL wrapper on saved objects client, so if user is trying to call the saved objects API, it is still under our ACL permission control.
2. update API provided by OpenSearch: Overall, we can not stop user from doing this. But in the future, few users will have permission(security plugin provides) to update `.kibana` index.

File: src/plugins/workspace/server/utils.ts
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/utils.ts:11]: For generation method of workspace id, we have a [RFC](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5201) to discuss.

For workspace id in the url path, we do compare some options in the [RFC](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5243) we can use and in the end, prepending that into OSD path seems more reasonable. But this topic should not be discussed in this PR.

**Review by zengyan-amazon — APPROVED** [review]

**Review by joshuarrrr — APPROVED** [review]

**Comment by kavilla** [comment]
Based off of https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5409#issuecomment-1789590501 I think this is okay its not in 2.x yet but the backport might be difficult in the future.

@SuZhou-Joe @ruanyl @xluo-aws do you have information to share about status of workspace feature.

And is it okay that I removed the 2.12 label?

**Comment by xluo-aws** [comment]
Rocky, thanks for the reminder, the new target release is 2.14.

**Pull request merged**
